### PR TITLE
Keep each target profile's share of the binomial STC gradient

### PR DESCRIPTION
**Pull request by choxos**
## What was wrong

A target made of many copies of one covariate profile is the same point mass
however many copies it holds, so standardizing it returns the point and the
link-scale derivative in an intercept is exactly 1. Under the complementary
log-log link it was not:

| eta | copies | link gradient | correct |
|---:|---:|---:|---:|
| 35 | 64 | 0.913 | 1 |
| 38 | 16 | 0.293 | 1 |
| 38 | 128 | 2.344 | 1 |
| 40 | 64 | 64 | 1 |

For a one-coordinate covariance this multiplies the standard error by the
number of grid points and the variance by its square.

## Why

The complementary log-log non-event log probability is `-exp(eta)`, which is
`-2.4e17` at eta 40. A double's neighbors are 32 apart there, so adding a log
weight of order 1 to it changes nothing: every point's normalized share came
back as 1 instead of 1/n and the grid resolution entered the answer. At
eta 38 the loss is partial and the factor is whatever the rounding left.

Two sites added the weight before subtracting the common term. Both now
subtract first, which is exact for values that close together, and add the
weight to a number of order 1.

* `.weighted_log_mean_exp()` shifts by the largest value before the weights
  are added rather than after, so the mean of identical values is exactly
  that value however they are weighted.
* `.stc_binomial_gradients()` forms each share as
  `exp(log_w + (log_p - log_p_mean))`.

While here, the complementary log-log link gradient switched to its
small-probability form at `log p-bar < -18`, where the exact form
`d log q-bar / log q-bar` is still available: the non-event log probability
is built as `-exp(eta)`, not as `log(1 - p)`, so it stays representable until
`exp(eta)` itself underflows below eta -745. The switch is now on that, which
is what the documentation already described, and the point-mass derivative at
eta -20 is 1 rather than `1 - p-bar / 2`.

## Tests

`test-stc-binomial-gradients.R` gains three cases: the replicated point mass
for all three links across eta in {-40, -20, 0, 20, 35, 38, 40} and 1 to 128
copies; unequal weights on identical profiles and a rescaling of all weights;
and that both sets of normalized shares sum to one. The known answer is 1,
not merely a finite number, which is what the existing tail tests checked.

Full suite: 4315 passing, 0 failures, 0 errors.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved numerical stability for weighted calculations involving extremely large or small values.
  * Fixed binomial gradient calculations to preserve very small contributions and maintain accurate results with extreme predictors.
  * Improved handling of complementary log-log link gradients in edge cases.

* **Tests**
  * Added regression coverage across logit, probit, and complementary log-log links, including varied weights, profile replication, and extreme predictor values.

### CODE REVIEW

**Comment by choxos** [comment]
Fixed in 4d8c7c9. Confirmed first: `c(1e308, 1e308)` gave a link gradient of 0 where a point mass requires 1, under every link, because `sum(weights)` is `Inf` and `log_w` became `-Inf`. That is the same failure this branch repairs, reached by a different route.

The denominator is now a shifted log-sum-exp, matching how `.weighted_log_mean_exp()` normalizes its own. The case is pinned in `test-stc-binomial-gradients.R` beside the rescaling one, for all three links and four values of eta.

**Comment by choxos** [comment]
@coderabbitai full review

**Comment by choxos** [comment]
@coderabbitai full review

**Comment by choxos** [comment]
@codex review

**Comment by choxos** [comment]
@codex review

File: R/stc.R [hunk 1]
```diff
@@ -794,7 +809,7 @@ stc <- function(data, link = NULL, conf_level = 0.95, distribution = "weibull",
     # `.binary_link_from_logs()` reports is +Inf, and the gradient is NaN
     # with it; the finite-variance guard then refuses the fit, as it did
     # before, rather than attach a finite SE to an infinite estimate.
-    colSums(-exp(log_w + lp$nonevent - log_q_mean + eta) * X)
+    colSums(-exp(log_w + (lp$nonevent - log_q_mean) + eta) * X)
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/stc.R]: Confirmed and fixed in c762f11.

Reproduced through the package, and it is broader than the example: the link
gradient was 0.5 at every spacing, not only at 1e-14.

```
delta=0        link_grad=1.000000000000   (homogeneous, correct)
delta=1e-14    link_grad=0.500000000000
delta=1e-12    link_grad=0.500000000000
delta=1e-10    link_grad=0.500000000000
delta=1        link_grad=0.500000000000
delta=10       link_grad=0.500000000000
```

At eta = 40 beside eta = 50 the second point contributes nothing whatsoever
and the first still got 0.5, so the failure is not about near-ties. The
cause is as described: `log_p_mean` is the largest log probability plus a
correction of order one, and adding the correction back next to -2.4e17,
where a double's neighbors are 32 apart, loses it. The dominant point then
receives its own weight in place of the whole share.

The fix forms the shares in the frame where the maximum has already
cancelled and normalizes them there by their own log-sum-exp, so the
correction is never written beside the huge term and the shares sum to one
by construction:

```r
log_shares <- function(x) {
  m_x <- max(x)
  z <- (x - m_x) + log_w
  m_z <- max(z)
  z - (m_z + log(sum(exp(z - m_z))))
}
```

The cloglog non-event branch now exponentiates `log_share_q + eta` instead
of rebuilding its share too.

Checks:

* All spacings above now return 1, at weights `0.5/0.5`, `0.25/0.75` and
  `1/3`, and with the dominant point on either side.
* Not met by returning 1 everywhere: on targets where no point dominates the
  gradients are 0.9607422554 and 0.3529196214, and they agree with a central
  difference of the standardized link to 1e-10 wherever that difference
  resolves.
* The original case in this PR is unaffected: 64 copies of one profile still
  give 1 under all three links.
* Failing-first, with only `R/stc.R` reverted: 30 failures, and 0 after.
* Full pure-R suite: 66 files, 4371 passed, 0 failed, 0 errors. No lints.

One note on the existing coverage: the "normalized shares sum to one" test
asserted the old expression inline rather than calling the code, so it never
constrained the implementation and passed throughout this bug. It now builds
the shares the way the code does and includes two targets where the two
forms disagree.

**Comment by choxos** [comment]
@codex review

**Comment by choxos** [comment]
@coderabbitai full review

**Comment by choxos** [comment]
@coderabbitai full review
